flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocatorTest.java

```diff
@@ -0,0 +1,575 @@
+				return payload;
+			})
+			.collect(Collectors.toList());
+		TestingPhysicalSlot physicalSlot = context.getSlotProvider().getResponses().values().stream().findFirst().get().get();
+
+		assertThat(payloads.stream().allMatch(payload -> payload.getTerminalStateFuture().isDone()), is(false));
+		assertThat(physicalSlot.getPayload(), notNullValue());
+		physicalSlot.getPayload().release(new Throwable());
+		assertThat(payloads.stream().allMatch(payload -> payload.getTerminalStateFuture().isDone()), is(true));
+	}
+
+	private static List<ExecutionVertexID> getAssignIds(Collection<SlotExecutionVertexAssignment> assignments) {
+		return assignments.stream().map(SlotExecutionVertexAssignment::getExecutionVertexId).collect(Collectors.toList());
+	}
+
+	private static class AllocationContext {
+		private final List<ExecutionVertexSchedulingRequirements> requirements;
+		private final TestingPhysicalSlotProvider slotProvider;
+		private final TestingSlotSharingStrategy slotSharingStrategy;
+		private final SlotSharingExecutionSlotAllocator allocator;
+		private final TestingSharedSlotProfileRetrieverFactory slotProfileRetrieverFactory;
+
+		AllocationContext(
+				List<ExecutionVertexSchedulingRequirements> requirements,
+				TestingPhysicalSlotProvider slotProvider,
+				TestingSlotSharingStrategy slotSharingStrategy,
+				SlotSharingExecutionSlotAllocator allocator,
+				TestingSharedSlotProfileRetrieverFactory slotProfileRetrieverFactory) {
+			this.requirements = requirements;
+			this.slotProvider = slotProvider;
+			this.slotSharingStrategy = slotSharingStrategy;
+			this.allocator = allocator;
+			this.slotProfileRetrieverFactory = slotProfileRetrieverFactory;
+		}
+
+		public SlotSharingExecutionSlotAllocator getAllocator() {
+			return allocator;
+		}
+
+		private List<SlotExecutionVertexAssignment> allocateSlotsFor(int start, int end) {
+			return allocator.allocateSlotsFor(requirements.subList(start, end));
+		}
+
+		private TestingSlotSharingStrategy getSlotSharingStrategy() {
+			return slotSharingStrategy;
+		}
+
+		private List<ExecutionVertexID> getReqIds(int start, int end) {
+			return getReqIds(requirements.subList(start, end));
+		}
+
+		private TestingPhysicalSlotProvider getSlotProvider() {
+			return slotProvider;
+		}
+
+		private TestingSharedSlotProfileRetrieverFactory getSlotProfileRetrieverFactory() {
+			return slotProfileRetrieverFactory;
+		}
+
+		private static List<ExecutionVertexID> getReqIds(Collection<ExecutionVertexSchedulingRequirements> requirements) {
+			return requirements.stream().map(ExecutionVertexSchedulingRequirements::getExecutionVertexId).collect(Collectors.toList());
+		}
+
+		static Builder newBuilder() {
+			return new Builder();
+		}
+
+		private static class Builder {
+			private int[] groups = { 2, 1 }; // 2 executions in the first group, 1 in the second etc
+			private List<ResourceProfile> resourceProfiles;
```

Comment:
should we explicitly set default values for these fields?
NPE will be thrown if `resourceProfiles` is null when building the context. For the boolean fields, I feel that explicitly setting them to `false` by default can make it more readable.